x-pack/metricbeat/module/statsd/server/data_test.go

```diff
@@ -358,3 +358,65 @@ func TestCounter(t *testing.T) {
 		"metric01": map[string]interface{}{"count": int64(0)},
 	})
 }
+
+func TestCounterSampled(t *testing.T) {
+	ms := mbtest.NewMetricSet(t, map[string]interface{}{"module": "statsd"}).(*MetricSet)
+	testData := []string{
+		"metric01:1|c|@0.1",
+		"metric01:2|c|@0.2",
+	}
+	process(t, testData, ms)
+
+	events := ms.getEvents()
+	assert.Len(t, events, 1)
+
+	assert.Equal(t, events[0].MetricSetFields, common.MapStr{
+		"metric01": map[string]interface{}{"count": int64(20)},
+	})
+}
+
+func TestTimerSampled(t *testing.T) {
+	ms := mbtest.NewMetricSet(t, map[string]interface{}{"module": "statsd"}).(*MetricSet)
+	testData := []string{
+		"metric01:2|ms|@0.01",
+		"metric01:1|ms|@0.1",
+		"metric01:2|ms|@0.2",
+		"metric01:2|ms",
+	}
+
+	// total of 100 + 10 + 5 + 1 = 116 measurements
+	process(t, testData, ms)
+
+	// rate gorutine runs every 5 sec
+	time.Sleep(time.Second * 6)
+
+	events := ms.getEvents()
+	assert.Len(t, events, 1)
+
+	actualMetric01 := events[0].MetricSetFields["metric01"].(map[string]interface{})
+
+	// returns the extrapolated count
+	assert.Equal(t, int64(116), actualMetric01["count"])
+
+	// rate numbers are updated by a gorutine periodically, so we cant tell exactly what they should be here
+	// we just need to check that the sample rate was applied
```

Comment:
How does current code that the rate is applied?

Reply:
without the rate applied it would be calculating the rate based on a count of 4, one for each measurement. Instead, because the sampling rate is applied, the rate is calculated based of off a count of 116. Happy to change this is you have better ideas to assert it

Reply:
I guess that without mocking something it'd be difficult to make it more precise, let's go with it as is by now.